### PR TITLE
Mdetr pseudo labels

### PR DESCRIPTION
**Pull request by hanoonaR**
- Makes pretraining with MDETR and MDef-DETR proposals an option
- Add hubconfig
- Updated readme with instructions for using hubconfig

### CODE REVIEW

File: datasets/selfdet.py [hunk 1]
```diff
@@ -59,6 +62,12 @@ def __init__(self, root, detection_transform, query_transform, cache_dir=None, m
                     self.files.append(path)
                 else:
                     continue
+        if self.pseudo_labels in ['mdetr', 'mdef_detr']:
+            # Load mdetr or mdef_detr boxes
+            pseudo_dets_dets_file_path = f"{root}_{self.pseudo_labels}_dets.pkl"
```
Comment (mmaaz60) [inline datasets/selfdet.py:67]: Let's have a different `.pkl` file name for each model

File: README.md [hunk 2]
```diff
@@ -154,6 +154,19 @@ Training takes around 1.5 days with 8 NVIDIA V100 GPUs, you can download a pretr
...
+
+[comment]: <> (The MDETR proposals generated on ImageNet100 is saved in: )
+The command for pretraining DETReg on 8 GPUs on ImageNet100 is as following:
+- On MDef-DETR:
+```bash
+GPUS_PER_NODE=8 ./tools/run_dist_launch.sh 8 ./configs/DETReg_top30_in100.sh --pseudo_labels mdef_detr --batch_size 24 --num_workers 8
+```
+- On MDETR:
+```bash
+GPUS_PER_NODE=8 ./tools/run_dist_launch.sh 8 ./configs/DETReg_top30_in100.sh --pseudo_labels mdetr --batch_size 24 --num_workers 8
```
Comment (mmaaz60) [inline README.md]: Let's have a different training config (`.sh`) file for MDETR & MDef-DETR where we can specify different output directory names corresponding to the model.

File: README.md [hunk 3]
```diff
@@ -203,6 +215,25 @@ To evaluate a finetuned model, use the following command from the project basedi
...
+### Using hubconfig for evaluation:
+1. Pretrained model with MDef-DETR proposals:
+```bash
+checkpoints_path = "./saved_models/DETReg_top30_in100_mdef_detr.pth"
+model = deformable_detr_resnet50(pretrained=False, return_postprocessor=False, num_classes=91,
+                                 object_embedding_loss=True, checkpoints_path=checkpoints_path)
+```
+2. Fine tune on 10% of Pascal:
+```bash
+checkpoints_path = "./saved_models/DETReg_fine_tune_10pct_pascal_mdef_detr.pth"
```
Comment (mmaaz60) [inline README.md]: Let's make sure that the training configs that we are providing save the models with the same names as mentioned here

File: hubconf.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                                     obj_embedding_head='intermediate')
+    return deformable_detr
+
+
+def deformable_detr_resnet50(pretrained=False, return_postprocessor=False, num_classes=91, object_embedding_loss=False,
+                             checkpoints_path=""):
+    """
+    Deformable DETR R50 with 6 encoder and 6 decoder layers.
+    """
+    model = _make_deformable_detr("resnet50", dilation=False, num_classes=num_classes,
```
Comment (mmaaz60) [inline hubconf.py]: Let's change the function names from `deformable_detr` to `detreg`

File: main.py [hunk 5]
```diff
@@ -29,6 +29,8 @@
 from engine import evaluate, train_one_epoch, viz
 from models import build_model
 from models.backbone import build_swav_backbone, build_swav_backbone_old
+import ssl
+ssl._create_default_https_context = ssl._create_unverified_context
```
Comment (mmaaz60) [inline main.py]: We can add the same commands in the Deformable DETR repo as well, or otherwise remove it from here. Both are fine

File: main.py [hunk 6]
```diff
@@ -289,21 +294,21 @@ def match_name_keywords(n, name_keywords):
...
-                                              data_loader_val, base_ds, device, args.output_dir)
-        if args.output_dir:
-            utils.save_on_master(coco_evaluator.coco_eval["bbox"].eval, output_dir / "eval.pth")
-        return
+    # if args.eval:
+    #     test_stats, coco_evaluator = evaluate(model, criterion, postprocessors,
+    #                                           data_loader_val, base_ds, device, args.output_dir)
+    #     if args.output_dir:
+    #         utils.save_on_master(coco_evaluator.coco_eval["bbox"].eval, output_dir / "eval.pth")
+    #     return
```
Comment (mmaaz60) [inline main.py]: The evaluation code was commented out to increase the overall training speed. It will be required in the final code for the evaluation of VOC-trained models.

Comment (mmaaz60) [reply main.py]: We will be uncommented the evaluation code at other places in this file as well.

**Review by mmaaz60 — COMMENTED** [review]
Let's review the added comments and merge the pull request

File: datasets/selfdet.py (hunk 1 above)
Comment (hanoonaR) [reply datasets/selfdet.py:67]: The pickle files will be named, `train_mdetr_dets.pkl` and `train_mdef_detr_dets.pkl`.

File: hubconf.py (hunk 4 above)
Comment (hanoonaR) [reply hubconf.py]: Changing only `deformable_detr_resnet50` to  `detreg_resnet50`.

File: README.md [hunk 7]
```diff
@@ -188,8 +189,42 @@ Fine tune on 10% of Pascal:
 GPUS_PER_NODE=2 ./tools/run_dist_launch.sh 2 ./configs/DETReg_fine_tune_10pct_pascal.sh --batch_size 4 --epochs 200 --lr_drop 150
 ```
 
+### Pretraining on ImageNet with pseudo labels from MDETR / MDef-DETR proposals
 
+[comment]: <> (The MDETR proposals generated on ImageNet100 is saved in: )
```
Comment (mmaaz60) [inline README.md:194]: I didn't understand this line, what is the meaning of it and why it is here?

File: README.md [hunk 8]
```diff
@@ -188,8 +189,42 @@ Fine tune on 10% of Pascal:
...
+The command for pretraining DETReg on 8 GPUs on ImageNet100 is as following:
```
Comment (mmaaz60) [inline README.md:195]: Let's change the commands to `training on 4 gpus` instead as we trained all DETReg models on 4 GPUs

File: README.md [hunk 9]
```diff
@@ -203,6 +238,45 @@ To evaluate a finetuned model, use the following command from the project basedi
...
+checkpoints_path = "./exps/DETReg_mdetr_prop_fine_tune_10pct_pascal/checkpoint.pth"
+model = deformable_detr_resnet50(pretrained=False, return_postprocessor=False, num_classes=21,
+                                 object_embedding_loss=False, checkpoints_path=checkpoints_path)
+```
+3. Fine tune on full of Pascal:
+```bash
+checkpoints_path = "./exps/DETReg_mdetr_prop_fine_tune_full_pascal/checkpoint.pth"
+model = deformable_detr_resnet50(pretrained=False, return_postprocessor=False, num_classes=21,
+                                 object_embedding_loss=False, checkpoints_path=checkpoints_path)
+```
```
Comment (mmaaz60) [inline README.md:279]: I see for both MDETR & MDef-DETR, only the checkpoints path is different. If so, we can mention pointing `checkpoints_path` to the relevant model and avoid the duplicate instructions

**Review by mmaaz60 — CHANGES_REQUESTED** [review]
In addition to the added comments, we have to pass `filter_pct` argument to `VOCDetection` class at `line 379` in `main.py`. This was the mistake in the original DETReg repo. Please check

File: README.md (hunk 9 above)
Comment (hanoonaR) [reply README.md:279]: Yeah sure.
